main/eth.c

```diff
@@ -0,0 +1,168 @@
+#include "eth.h"
+#include <esp_err.h>
+#include <esp_eth.h>
+#include <esp_event_loop.h>
+#include <esp_log.h>
+#include <mdns.h>
+#include <eth_phy/phy_lan8720.h>
+#include <eth_phy/phy_tlk110.h>
+#include <eth_phy/phy_ip101.h>
+#include <tcpip_adapter.h>
+#include <arpa/inet.h>
+#include <stdbool.h>
+#include <string.h>
+
+static const char *TAG = "Ethernet";
+
+static eth_on_connected_cb_t on_connected_cb = NULL;
+static eth_on_disconnected_cb_t on_disconnected_cb = NULL;
+static char *eth_hostname = NULL;
+static uint8_t eth_phy_gpio_power = 0;
+static eth_phy_t eth_phy;
+
+void eth_set_on_connected_cb(eth_on_connected_cb_t cb)
+{
+    on_connected_cb = cb;
+}
+
+void eth_set_on_disconnected_cb(eth_on_disconnected_cb_t cb)
+{
+    on_disconnected_cb = cb;
+}
+
+uint8_t *eth_mac_get(void)
+{
+    static uint8_t mac[6] = {};
+
+    if (!mac[0])
+        esp_read_mac(mac, ESP_MAC_ETH);
+
+    return mac;
+}
+
+void eth_hostname_set(const char *hostname)
+{
+    if (eth_hostname)
+        free(eth_hostname);
+    eth_hostname = strdup(hostname);
+}
+
+
+
+static esp_err_t event_handler(void *ctx, system_event_t *event)
+{
+    switch(event->event_id) {
+    case SYSTEM_EVENT_ETH_START:
+        if (eth_hostname)
+            tcpip_adapter_set_hostname(TCPIP_ADAPTER_IF_ETH, eth_hostname);
+        break;
+    case SYSTEM_EVENT_ETH_GOT_IP:
+        ESP_LOGD(TAG, "Got IP address: %s",
+            inet_ntoa(event->event_info.got_ip.ip_info.ip));
+        if (on_connected_cb)
+            on_connected_cb();
+        break;
+    case SYSTEM_EVENT_ETH_CONNECTED:
+        ESP_LOGI(TAG, "Connected");
+        break;
+    case SYSTEM_EVENT_ETH_DISCONNECTED:
+        ESP_LOGI(TAG, "Disconnected");
+        if (on_disconnected_cb)
+            on_disconnected_cb();
+        break;
+    default:
+        ESP_LOGD(TAG, "Unhandled event (%d)", event->event_id);
+        break;
+    }
+
+    mdns_handle_system_event(ctx, event);
+    return ESP_OK;
+}
+
+eth_phy_t eth_phy_atophy(const char *phy)
+{
+    struct {
+        const char *name;
+        int phy;
+    } *p, phys[] = {
+        { "LAN8720", PHY_LAN8720 },
+        { "TLK110", PHY_TLK110 },
+        { "IP101", PHY_IP101 }
+    };
+
+    for (p = phys; p->name; p++)
+    {
+        if (!strcmp(p->name, phy))
+            break;
+    }
+
+    ESP_LOGI(TAG, "PHY is %s", p->name);
```

Comment:
If we want to print the currently used Ethernet PHY, I would put this log message in `eth_connect()` and not here.